aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move

```diff
@@ -1,26 +1,118 @@
+/// underflow or on overflowing a custom upper limit). The difference from integers
+/// is that aggregators allow to perform both additions and subtractions in parallel
+/// across multiple transactions, enabling parallel execution. For example, if the
+/// first transaction is doing `try_add(X, 1)` for aggregator `X`, and the second is
+/// doing `try_sub(X,3)`, they can be executed in parallel avoiding a read-modify-write
+/// dependency.
+/// However, reading the aggregator value (i.e. calling `read(X)`) is an expensive
+/// operation and should be avoided as much as possible because it reduces the
+/// parallelism.
 module aptos_framework::aggregator_v2 {
+    use std::error;
     use std::string::String;
 
+    /// The value of aggregator overflows. Raised by uncoditional add() call
+    const EAGGREGATOR_OVERFLOW: u64 = 1;
+
+    /// The value of aggregator underflows (goes below zero). Raised by uncoditional sub() call
+    const EAGGREGATOR_UNDERFLOW: u64 = 2;
+
     /// The generic type supplied to the aggregator snapshot is not supported.
     const EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE: u64 = 5;
 
     /// The aggregator snapshots feature flag is not enabled.
     const EAGGREGATOR_SNAPSHOTS_NOT_ENABLED: u64 = 6;
 
+    /// The generic type supplied to the aggregator is not supported.
+    const EUNSUPPORTED_AGGREGATOR_TYPE: u64 = 7;
+
+    /// Arguments passed to concat exceed max limit of 256 bytes (for prefix and suffix together)
+    const ECONCAT_STRING_LENGTH_TOO_LARGE: u64 = 8;
+
+    /// The native aggregator function, that is in the move file, is not yet supported.
+    /// and any calls will raise this error.
+    const EAGGREGATOR_FUNCTION_NOT_YET_SUPPORTED: u64 = 9;
+
+    /// Represents an integer which supports parallel additions and subtractions
+    /// across multiple transactions. See the module description for more details.
+    ///
+    /// Currently supported types for Element are u64 and u128.
+    struct Aggregator<IntElement> has store {
+        value: IntElement,
+        max_value: IntElement,
+    }
+
+    /// Represents a constant value, that was derived from an aggregator at given instant in time.
+    /// Unlike read() and storing the value directly, this enables parallel execution of transactions,
+    /// while storing snapshot of aggregator state elsewhere.
     struct AggregatorSnapshot<Element> has store, drop {
         value: Element,
     }
 
+    /// Returns `max_value` exceeding which aggregator overflows.
+    public fun max_value<IntElement: copy + drop>(aggregator: &Aggregator<IntElement>): IntElement {
+        aggregator.max_value
+    }
+
+    /// Creates new aggregator, with given 'max_value'.
+    ///
+    /// Currently supported types for Element are u64 and u128.
+    /// EAGGREGATOR_ELEMENT_TYPE_NOT_SUPPORTED raised if called with a different type.
+    public native fun create_aggregator<IntElement: copy + drop>(max_value: IntElement): Aggregator<IntElement>;
```

Comment:
if we only support u64 and u128, I think it's better to write constructors explicitly and leave other functions generic
```
  fun create_u64_aggregator(max_value: u64): Aggregator<u64>;
```

Reply:
why? 

usage here seems just fine - https://github.com/aptos-labs/aptos-core/pull/9972/files#diff-7026c328622536165825a6f5dcaf13764bd475489ecba30ba4716dfdc4acc1e9

and set of types we support is something we can expand on in the future.